src/cts/src/HTreeBuilder.cpp

```diff
@@ -224,6 +225,456 @@ void HTreeBuilder::initSinkRegion()
+    const std::vector<Point<double>>& legal_locations,
+    const std::vector<Point<double>>& sinks)
+{
+  Point<double> ans = legal_locations.front();
+  double minDist = weightedDistance(ans, original_loc, sinks);
+  for (const Point<double>& x : legal_locations) {
+    double d = weightedDistance(x, original_loc, sinks);
+    if (d < minDist) {  // choose one of legal_locations that is closest to
+                        // original_loc
+      minDist = d;
+      ans = x;
+    }
+  }
+  return ans;
+}
+
+void plotSinks(std::ofstream& file, const std::vector<Point<double>>& sinks)
+{
+  unsigned cnt = 0;
+  for (const Point<double>& pt : sinks) {
+    double x = pt.getX();
+    double y = pt.getY();
+    double w = 1;
+    double h = 1;
+    auto name = "sink_";
+    file << cnt++ << " " << x << " " << y << " " << w << " " << h;
+    file << " " << name << " " << std::endl;
+  }
+}
+
+unsigned HTreeBuilder::findSibling(LevelTopology& topology,
+                                   unsigned i,
+                                   unsigned par)
+{
+  unsigned idx = 0;
+  for (; idx < topology.getBranchingPointSize(); ++idx) {
+    unsigned k = topology.getBranchingPointParentIdx(idx);
+    if (idx != i && k == par) {
+      return idx;
+    }
+  }
+  return i;
+}
+
+void scalePosition(Point<double>& loc,
+                   const Point<double>& parLoc,
+                   double leng,
+                   double scale)
+{
+  double px = parLoc.getX();
+  double py = parLoc.getY();
+  double ax = loc.getX();
+  double ay = loc.getY();
+
+  double d = loc.computeDist(parLoc);
+  double x, y;
+  if (d > 0) {  // yy8
+    double delta = d * scale;
+    double dx = ax - px;
+    double dy = ay - py;
+    dx += (dx > 0) ? delta : -delta;
+    dy += (dy > 0) ? -delta : delta;
+    double scale = leng / d;
+    x = px + dx * scale;
+    y = py + dy * scale;
+  } else {
+    x = px + leng / 2;
+    y = py + leng / 2;
+  }
+  loc.setX(x);
+  loc.setY(y);
+}
+
+void setSiblingPosition(const Point<double>& a,
+                        Point<double>& b,
+                        const Point<double>& parLoc)
+{
+  double px = parLoc.getX();
+  double py = parLoc.getY();
+  double ax = a.getX();
+  double ay = a.getY();
+  double bx = 2 * px - ax;
+  double by = 2 * py - ay;
+  b.setX(bx);
+  b.setY(by);
+}
+
+/* xxx Balance the two branches on the very top level */
+void adjustToplevelTopology(Point<double>& a,
+                            Point<double>& b,
+                            const Point<double>& parLoc)
+{
+  double da = a.computeDist(parLoc);
+  double db = b.computeDist(parLoc);
+  if (da < db) {
+    setSiblingPosition(a, b, parLoc);
+  } else if (db < da) {
+    setSiblingPosition(b, a, parLoc);
+  }
+}
+
+bool moveOnBlockBounardy(const Point<double>& pt,
+                         Point<double>& qt,
+                         double& x1,
+                         double& y1,
+                         double& x2,
+                         double& y2)
```

Comment:
Why are these taken by non-const ref?  The don't look to be modified in this method.

Reply:
Removed reference